src/compare-tree/index.js

```diff
@@ -0,0 +1,197 @@
+import Node from './node'
+import Query from './query'
+
+import {
+  castPath,
+  getKeyPaths,
+  getKeyString,
+  getKeyStrings
+} from '../key-path'
+
+const ROOT_KEY = '~'
+
+export default function CompareTree (initial) {
+  this.snapshot = initial
+  this.nodes = {}
+}
+
+CompareTree.prototype = {
+
+  /**
+   * Create a subscription to a particular set of key paths.
+   * @public
+   * @param {String} keyPaths A list of key paths to trigger the callback
+   * @param {Function} callback Function to execute when a key path changes
+   * @param {*} [scope] Optional scope to invoke the function with
+   */
+  on (keyPaths, callback, scope) {
+    let dependencies = getKeyPaths(keyPaths)
+    let id = Query.getId(keyPaths)
+
+    let query = this.addQuery(id, dependencies)
+
+    dependencies.map(this.track, this)
+                .forEach(node => node.connect(query))
+
+    query.on('change', callback, scope)
+
+    return query
+  },
+
+  /**
+   * Remove a subscription created by .on()
+   * @public
+   * @param {String} keyPaths A list of key paths for the subscription
+   * @param {Function} callback Associated callback function
+   * @param {*} [scope] Associated scope
+   */
+  off (keyPaths, callback, scope) {
+    let dependencies = getKeyPaths(keyPaths)
+    let id = Query.getId(keyPaths)
+
+    let query = this.nodes[id]
+
+    if (query) {
+      query.off('change', callback, scope)
+
+      if (query._events <= 0) {
+        this.prune(query, dependencies)
+      }
+    }
+  },
+
+  /**
+   * Compare a new snapshot to the last snapshot, triggering event
+   * subscriptions along the way.
+   * @public
+   * @param {*} state - New snapshot of state
+   */
+  update (snapshot) {
+    let last = this.snapshot
+    this.snapshot = snapshot
+
+    let root = this.nodes[ROOT_KEY]
+
+    if (root) {
+      this.scan(root, last, snapshot)
+    }
+  },
+
+  /**
+   * Add a node to the tree if it has not otherwise been added.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} key Each node represents a key in a nested object.
+   */
+  addNode (id, key) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Node(id, key)
+    }
+
+    return this.nodes[id]
+  },
+
+  /**
+   * Add a query to the tree if it has not otherwise been
+   * added. Queries are leaf nodes responsible for managing
+   * subscriptions.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} keyPaths Each query tracks a list of key paths
+   */
+  addQuery (id, keyPaths) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Query(id, keyPaths)
+    }
+
+    return this.nodes[id]
+  },
```

Comment:
I split up `add` into `addQuery` and `addNode`. I like this better, but wonder if we could just have a single node class. 

The trick is that nodes represent a single key name in a JS object, whereas Queries represent multiple pathways.